Suggestions to optimize `Document` iteration

joshfeli opened the pull request:
# Suggestions to optimize `Document` iteration

## Overview
This PR offers a solution to Issue #20 by presenting an example optimization in `app.analysis.__init__.py`. **This is not to be merged into the `main` branch:** this code is just here for reference for anyone working through similar optimization issues.

The optimization here takes advantage of Django `QuerySet`s' [lazy evaluation](https://docs.djangoproject.com/en/3.1/topics/db/queries/#querysets-are-lazy) and makes many database queries (as opposed to a bulk query used for iteration) to delay hitting the database until absolutely needed. We use a `Document`'s primary key (the `pk` field in the database table) to filter a `QuerySet` that returns a list of field values (rather than entire `Document` instances) to limit what we load from the database, achieved using the `QuerySet`'s `.values_list()` and `.filter()` methods.

Here's some [documentation](https://docs.djangoproject.com/en/3.1/ref/models/querysets/) for more details.

joshfeli committed:
Copied and pasted code from GATK proximity.py

joshfeli committed:
Labeled a starting point for working with proximity.py

joshfeli committed:
Merge branch 'main' into transfer_proximity_analysis

farooqashar committed:
Updated _generate_token_counter helper function

Updated the _generate_token_counter helper function to reflect the changes of the document and gender models

Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

farooqashar committed:
Updated reference to the new label of a document model

Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

farooqashar committed:
Efficient looping and results modification revisions

Implemented an efficient way of looping through documents and revised the results dictionary for various analysis functions

Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Reformatted spacing in `proximity.py`

joshfeli committed:
Merge branch 'main' into transfer_proximity_analysis

farooqashar committed:
Added fixtures to add initial data to the pronoun series model

Added json files with initial data

Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>

joshfeli committed:
Optimized performance iterating through `Document`s

Refactored iteration through `Document`s of a `Corpus` such that Django only pulls the relevant fields of each instance from the database rather than each complete `Document` object. We optimized this performance only in `proximity.py`.

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Revert "Added fixtures to add initial data to the pronoun series model"

This reverts commit 04adffcb974c0c40588dbe07df326310d522f215.

MBJean reviewed (changes requested):
Great start on this. I have a few high-level comments to start.
- Would it be possible to reduce this PR to just `_run_analysis` and the function(s) called in it? That'll help us focus in in this PR on exactly the goals of the PR (finding the most optimized way to iterate through these documents). If you do that, you should keep your work on the other methods and functions in this PR. You can circle back to your work once you've got a final draft of `_run_analysis` going.
- In line 349, do we know how calling `self.corpus.documents.values` will work? Will it pull all `document` `tokenized_text` values into memory, or just those required for that cycle of the iteration. I would assume the former, though Python's ORM might be doing something fancy behind the scene there. I suggest reading through https://docs.djangoproject.com/en/3.2/topics/db/optimization/ to figure out the best strategy to minimize memory use (even if at the cost of speed).
- We should plan to dispense with some of the features of this part of the code that were important when this was architected as a standard Python package. I'm thinking we can get rid of type hinting, a bunch of the imports, and the class structure for `GenderProximityAnalyzer`.

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Isolated `run_analysis`; added mechanisms for creating `Document` objects

Import `app.analysis` in your Django shell and call `make_documents(num)` to create the `Document` objects, and call `run_analysis()` with a `QuerySet` [of `Document`s] passed in.

joshfeli committed:
Refined docstring of isolated `run_analysis` function in `__init__.py` in the `analysis` module

joshfeli committed:
Added a variable binding for the query

joshfeli committed:
Trying out a third-party memory profile and server-side cursors in SQL

I'm attempting here to optimize `Document` iteration by implementing a server-side cursor using raw SQL. I'm also using the third-party `memory_profiler` Python package to track any differences in memory usage with this approach. So far, my `DECLARE` query raises a `SyntaxError`.

joshfeli committed:
Added comments to highlight error

joshfeli committed:
Updated comments in `__init__.py` of `analysis` module

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Added `psutil` to `requirements.txt`

The `psutil` package isn't used directly, but it is a dependency for the `memory-profiler` package.

joshfeli committed:
Continuing to experiment with memory profiling

(WORK IN PROGRESS!)
I've added the `eliot_middlemarch.txt` file and have used it instead of `aarund_longfrock.txt` since the former is 1.7MB while the latter is only a couple of KB. I'm using the `@profile` decorator alongside printing the length of the `connection.queries` list after each line to check where/when I'm hitting the DB as well  as how much memory each hit costs.

joshfeli committed:
Corrected `make_documents` so each `Document`'s text is tokenized

This is a crucial correction since the `tokenized_text` field in each Model instance was empty until now!

joshfeli committed:
Added and corrected docstrings to functions in `app/analysis/__init__.py`

joshfeli committed:
Modified docstrings further in `app/analysis/__init__.py`

joshfeli committed:
Trying another approach to optimization

Adding more `Document`s, trying a different approach of querying DB (extreme case of making a new SQL query for each `Document`.

This supposedly is not best practice, but I'm trying it out to see if there are any memory spikes.

I'm also playing with the `gc` built-in module to track [garbage collection](https://docs.python.org/3/library/gc.html).

joshfeli committed:
Mimicking a situation that will require `Document` iteration

Iterating through `Document`s in a "global" frame (i..e in the `analysis_wrapper` function), using each `Document` instance for analysis and calling an analysis function for each instance, returning the results in the "global" frame.

joshfeli committed:
Added some comments to help explain the optimization

joshfeli committed:
Implemented `Document` iteration optimization in `proximity.py`

Optimized iteration through `Document` objects as described in the demos in the `__init__.py` file of the `analysis` module

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Added a description to the `FieldError` `except` block

joshfeli committed:
Added some extra comments to the `Document` iteration demo

Credit to @Watermelanie for the idea of only passing down the primary key!

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Removed all changes to `proximity.py`

(Left for another PR)

joshfeli committed:
Commented out old ideas for optimization

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Corrected a typo

joshfeli committed:
Updated comments

MBJean commented:
@joshfeli Since we've worked through this challenge and merged the `ProximityAnalysis` project into `main`, I think we're free to close this PR. That seem legit to you?

joshfeli commented:
> @joshfeli Since we've worked through this challenge and merged the `ProximityAnalysis` project into `main`, I think we're free to close this PR. That seem legit to you?

Sounds good! Closing now.
